### PR TITLE
fMP: Use last fMPCandidate if no FMP is marked.

### LINKED ISSUE #1564: FMP: Improve reliability on long pageloads
The [FirstMeaningfulPaintDetector.cpp](https://cs.chromium.org/chromium/src/third_party/WebKit/Source/core/paint/FirstMeaningfulPaintDetector.cpp?sq=package:chromium) has a network idle detection that's a little more conservative than ours. As such, it will sometimes wait longer than we will to choose an FMP.

In this cases, we stop the trace and look in it for a particular "firstMeaningfulPaint" but it wasn't emitted yet. 

what we should do instead:
* look at the "firstMeaningfulPaintCandidate" trace events from that thread & frame. 
* Choose the last one.
* Consider that the FMP.

(basically, the same thing the native impl does)

### PR DESCRIPTION
**Pull request by paulirish**
Good test page for this is https://madeby.google.com/phone/ which we always time out after 30s

fixes #1564 

--

fmpCandidate comes from `loading` trace category (for whatever reason). 

The changes in pwmetrics are for handling -1 failures a little better.

### CODE REVIEW

File: lighthouse-core/audits/first-meaningful-paint.js [hunk 1]
```diff
@@ -53,9 +54,19 @@ class FirstMeaningfulPaint extends Audit {
...
       if (!tabTrace.firstMeaningfulPaintEvt) {
-        throw new Error('No usable `firstMeaningfulPaint` event found in trace');
+        const fmpCand = 'firstMeaningfulPaintCandidate';
+        log.verbose('fmp-audit', `No firstMeaningfulPaint found, falling back to last ${fmpCand}`);
+        const candidates = tabTrace.processEvents.filter(e => e.name === fmpCand);
+
+        if (candidates.length === 0) {
+          throw new Error('No usable `firstMeaningfulPaint(Candidate)` events found in trace');
+        }
+        tabTrace.firstMeaningfulPaintEvt = candidates.pop();
```
Comment (patrickhulce) [inline lighthouse-core/audits/first-meaningful-paint.js]: do you think any other audits would need to use this inferred fmp event? or is this it?

File: lighthouse-core/lib/traces/pwmetrics-events.js [hunk 2]
```diff
@@ -148,7 +148,7 @@ class Metrics {
    * @param {number} navStartTs
```
Comment (patrickhulce) [inline lighthouse-core/lib/traces/pwmetrics-events.js]: update this?

File: lighthouse-core/lib/traces/pwmetrics-events.js [hunk 3]
```diff
@@ -182,8 +182,8 @@ class Metrics {
     }
 
     // confirm our navStart's correctly match
-    const navStartTs = metrics.find(e => e.id === 'navstart').ts;
-    if (this.getNavigationStartEvt().ts !== navStartTs) {
+    const navStartEvt = metrics.find(e => e.id === 'navstart');
+    if (!navStartEvt || this.getNavigationStartEvt().ts !== navStartEvt.ts) {
```
Comment (patrickhulce) [inline lighthouse-core/lib/traces/pwmetrics-events.js]: does `this.getNavigationStartEvt()` always find something or is this another candidate for TypeError too?

Comment (paulirish) [inline lighthouse-core/lib/traces/pwmetrics-events.js]: yup it should always find something.

The `getNavigationStartEvt` method could live on top of trace-of-tab as well, though...

Comment (paulirish) [inline lighthouse-core/lib/traces/pwmetrics-events.js]: ^ filed that over here https://github.com/GoogleChrome/lighthouse/issues/1568

**Comment by paulirish** [comment]
PTAL

File: lighthouse-core/gather/computed/trace-of-tab.js [hunk 4]
```diff
@@ -67,9 +69,23 @@ class TraceOfTab extends ComputedArtifact {
...
+    // If there was no firstMeaningfulPaint event found in the trace, the network idle detection
+    // may have not been triggered before Lighthouse finished tracing.
+    // In this case, we'll use the last firstMeaningfulPaintCandidate we can find.
+    // However, if no candidates were found (a bogus trace, likely), we fail.
+    if (!firstMeaningfulPaint) {
+      const fmpCand = 'firstMeaningfulPaintCandidate';
+      log.verbose('trace-of-tab', `No firstMeaningfulPaint found, falling back to last ${fmpCand}`);
+      const lastCandidate = frameEvents.filter(e => e.name === fmpCand).pop();
+      if (!lastCandidate) {
+        log.verbose('No `firstMeaningfulPaintCandidate` events found in trace');
```
Comment (brendankenny) [inline lighthouse-core/gather/computed/trace-of-tab.js]: 'trace-of-tab' for log title?

Comment (paulirish) [inline lighthouse-core/gather/computed/trace-of-tab.js]: done

**Comment by paulirish** [comment]
Travis is confused, but the tests are in good shape. :)

**Pull request merged**